### PR TITLE
[eclipse/xtext-xtend#1061] Provide ui test infrastructure for testing (WIP)

### PR DESCRIPTION
**Pull request by drkstr101**
multi quickfixes [2/3]

Added `EqualsWithNullMultiQuickfixTest` and
`RemoveUnnecessaryModifiersMultiQuickfixTest`

Signed-off-by: Aaron R Miller <aaron.miller@waweb.io>

### CODE REVIEW

**Comment by drkstr101** [comment]
TODO RemoveUnnecessaryModifiersMultiQuickfixTest contains failing test cases

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+				new Runnable {
+					public override def run() {
+						println(A)
+					}
+				}
+			}
+		}
+	'''
+
+	XtextEditor xtextEditor
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:43]: Move the `xtextEditior` variable into the `testRemoveUnnecessaryModifiersQuickfixJava` method

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+							println(A)
+						}
+					}
+				}
+			}
+			----------------------------------------------------
+			0: message=The public modifier is unnecessary on class Foo
+			1: message=The private modifier is unnecessary on field A
+			2: message=The final modifier is unnecessary on field A
+			3: message=The value of the field Foo.A is not used
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:82]: Remove `3: message=The value of the field Foo.A is not used` from the expected values.

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+	@Test
+	def void testRemoveUnnecessaryModifiersQuickfixJava() {
+
+		// Use Java style API
+		// //
+		
+		xtextEditor = openEditor(dslFile(MODEL_WITH_UNNECESSARY_MODIFIERS))
+
+		val offset = MODEL_WITH_UNNECESSARY_MODIFIERS.indexOf("==") + 1
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:111]: Replace `==` by `public`.

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		val proposals = computeQuickAssistProposals(xtextEditor, offset)
+
+		// TODO java.lang.AssertionError: expected:<1> but was:<0>
+		assertEquals(1, proposals.size())
+		assertEquals(1, proposals.filter[it instanceof QuickAssistCompletionProposal].size())
+
+		// TODO assert proposal text...
```
Comment (miklossy) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:118]: Add an assertion here, e.g. `assertEquals("Remove the unnecessary modifier.", proposals.head.displayString)`

**Comment by miklossy** [comment]
> TODO RemoveUnnecessaryModifiersMultiQuickfixTest contains failing test cases

It is actually quite strange for me also why the same setup works for the `EqualsWithNullMultiQuickfixTest` but not for the `RemoveUnnecessaryModifiersMultiQuickfixTest`.

With the following workaround the `RemoveUnnecessaryModifiersMultiQuickfixTest` succeed as well:


import org.eclipse.core.resources.IFile
import org.eclipse.core.resources.IMarker
import org.eclipse.core.runtime.NullProgressMonitor
...

class RemoveUnnecessaryModifiersMultiQuickfixTest extends AbstractMultiQuickfixTest {
	...
	// TODO: verify why this workaround is necessary
	override protected IMarker[] getMarkers(IFile file) {
		saveAllOpenEditors
		super.getMarkers(file)
	}

	private def saveAllOpenEditors() {
		activePage.editorReferences.forEach[
			getEditor(false).doSave(new NullProgressMonitor)
		]
	}
}

**Comment by drkstr101** [comment]
@miklossy Thanks for the input! I will make the changes you outlined in your comments and resubmit.

Cheers!

**Comment by miklossy** [comment]
@drkstr101 How is the progress here?

**Comment by drkstr101** [comment]
@miklossy I apologize, I meant to leave a comment earlier. I am having trouble getting the input strings just right in `RemoveUnnecessaryModifiersMultiQuickfixTest`. The markers seem to be correct, I believe my `---` separator may be the incorrect length. I am about to take another stab at it now.

File: org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+		projectName.createPluginProject()
+	}
+
+	@Test
+	def void testRemoveUnnecessaryModifiersQuickfixXtend() {
+
+		// Use Xtend style API...
+		
+		// TODO...
```
Comment (drkstr101) [inline org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:61]: @miklossy here is the error output

Comment (miklossy) [reply org.eclipse.xtend.ide.tests/longrunning/src/org/eclipse/xtend/ide/tests/quickfix/RemoveUnnecessaryModifiersMultiQuickfixTest.xtend:61]: @drkstr101 : In order to make the `RemoveUnnecessaryModifiersMultiQuickfixTest` test cases succeed, you have to ensure that the editors are saved before calculating the markers, see my previous [comment](https://github.com/eclipse/xtext-xtend/pull/1085#issuecomment-657361334) Could you please add these two methods (`getMarkers`, `saveAllOpenEditors`) to your PR?

**Comment by miklossy** [comment]
@drkstr101 : Thanks for your contribution Aaron! I am merging this PR right now and will continue to complete this issue.

**Comment by miklossy** [comment]
@drkstr101 I created a follow up PR: https://github.com/eclipse/xtext-xtend/pull/1090

**Comment by drkstr101** [comment]
@miklossy @cdietrich 

Sorry you had to pick up a bit of the slack for me here. It looks like everything is taken care of now, but I still have a few notifications I need to touch. I do have some more free time this week if you have anything else you would like me to take a look at.

I also have some feedback on `AbstractMultiQuickfixTest` api if you are interested. My only comment is that I found it somewhat difficult to guess at the expected input format of the input strings. I had to do a lot of trial and error fiddling, and it never was fully clear what the last issue was since the important text was truncated in the output. It is also possible I was just missing something obvious in the assertion error. 

Coming up with a better solution is probably still a bit over my pay grade at this point however. I am starting to learn more of the internals though, if there was anything else you would like me to dig into. I can commit to roughly 30 hours or so over the course of this week(end).

**Comment by miklossy** [comment]
@drkstr101 : Thanks for further offering your support! We think that https://github.com/eclipse/xtext-core/issues/897 would be a valuable contribution to the Xtext framework and we would be happy if you could take a look at that.

**Comment by cdietrich** [comment]
@drkstr101 feedback for AbstractMultiQuickfixTest is also welcome

**Comment by drkstr101** [comment]
@cdietrich My biggest complaint was requiring multiple runs to adjust my `-----`  line separators to get the desired length. Perhaps I need to adjust my font settings. This was amplified due to fiddling with the marker text which would then affect the line length and require more adjusting. I think it would be a fairly solid API if those line separators could be eliminated somehow.

**Comment by szarnekow** [comment]
I agree completely. The separators are cumbersome and don't provide added value.

**Comment by drkstr101** [comment]
Also worth noting, my time to implement the tests was effectively limited to the number of times I could launch the test workspace per hour, which was not very high on my machine unfortunately. This probably wouldn't be so bad if I wasn't depending on the test runs to discover what the inputs should be. `#badtester`

**Pull request merged**